src/Sarif.Viewer.VisualStudio/CodeAnalysisResultManager.cs

```diff
@@ -637,5 +611,220 @@ internal bool TryResolveFilePathFromSolution(string solutionPath, string pathFro
+            return false;
+        }
+
+        internal bool SaveResolvedPathToUriBaseMapping(string uriBaseId, string originalPath, string pathFromLogFile, string resolvedPath, RunDataCache dataCache)
+        {
+            Uri.TryCreate(pathFromLogFile, UriKind.Relative, out Uri relativeUri);
+            string fullPathFromLogFile = pathFromLogFile;
+            if (Uri.TryCreate(pathFromLogFile, UriKind.Absolute, out Uri absoluteUri))
+            {
+                fullPathFromLogFile = Path.GetFullPath(pathFromLogFile);
+            }
+            else
+            {
+                if (!fullPathFromLogFile.StartsWith("/"))
+                {
+                    fullPathFromLogFile = "/" + fullPathFromLogFile;
+                }
+            }
+
+            string commonSuffix = GetCommonSuffix(fullPathFromLogFile.Replace("/", @"\"), resolvedPath);
+            if (commonSuffix == null)
+            {
+                return false;
+            }
+
+            // Trim the common suffix from both paths, and add a remapping that converts
+            // one prefix to the other.
+            string originalPrefix = fullPathFromLogFile.Substring(0, fullPathFromLogFile.Length - commonSuffix.Length);
+            string resolvedPrefix = resolvedPath.Substring(0, resolvedPath.Length - commonSuffix.Length);
+
+            int uriBaseIdEndIndex = resolvedPath.IndexOf(originalPath.Replace("/", @"\"));
+
+            if (relativeUri != null && uriBaseIdEndIndex >= 0)
+            {
+                // If we could determine the uriBaseId substitution value, then add it to the map.
+                dataCache.RemappedUriBasePaths[uriBaseId] = new Uri(resolvedPath.Substring(0, uriBaseIdEndIndex), UriKind.Absolute);
+            }
+            else
+            {
+                // If there's no relativeUri/uriBaseId pair or we couldn't determine the uriBaseId value,
+                // map the original prefix to the new prefix.
+                dataCache.RemappedPathPrefixes.Add(new Tuple<string, string>(originalPrefix, resolvedPrefix));
+            }
+
+            return true;
+        }
+
+        // return false means cannot resolve local file and will use embedded file.
+        internal bool VerifyFileWithArtifactHash(SarifErrorListItem sarifErrorListItem, string pathFromLogFile, RunDataCache dataCache, string resolvedPath, string embeddedTempFilePath, out string newResolvedPath)
+        {
+            newResolvedPath = null;
+
+            if (string.IsNullOrEmpty(resolvedPath))
+            {
+                // cannot find corresponding file in local, then use embedded file
+                newResolvedPath = embeddedTempFilePath;
+                return true;
+            }
+
+            if (!dataCache.FileDetails.ContainsKey(pathFromLogFile))
+            {
+                newResolvedPath = resolvedPath;
+                return true;
+            }
+
+            ArtifactDetailsModel fileData = dataCache.FileDetails[pathFromLogFile];
```

Comment:
you could make something like:
if (!dataCache.FileDetails.TryGetValue(pathFromLogFile, out ArtifactDetailsModel fileData)
{
  newResolvedPath = resolvedPath;
  return true;
}

with that, you would not need the line 738

Reply:
This would also prevent you accessing the fileDetails two  times (1 for the contains and 1 getting the value)

---
In reply to: [606208141](https://github.com/microsoft/sarif-visualstudio-extension/pull/384#discussion_r606208141) [](ancestors = 606208141)

Reply:
fixed. also update other couple places using TryGetValue instead of check if contains and get value.

---
In reply to: [606208336](https://github.com/microsoft/sarif-visualstudio-extension/pull/384#discussion_r606208336) [](ancestors = 606208336,606208141)

Reply:
nice! well done!

---
In reply to: [606401848](https://github.com/microsoft/sarif-visualstudio-extension/pull/384#discussion_r606401848) [](ancestors = 606401848,606208336,606208141)